ingress/controller.go

```diff
@@ -0,0 +1,98 @@
+/*
+Package ingress contains the ingress controller which monitors Kubernetes
+and updates the ingress load balancer. It also runs the load balancer.
+ */
+package ingress
+
+import (
+	log "github.com/Sirupsen/logrus"
+	"github.com/sky-uk/feed/k8s"
+)
+
+// Controller for Kubernetes ingress.
+type Controller interface {
+	Run()
+	Stop()
+}
+
+type controller struct {
+	lb     LoadBalancer
+	client k8s.Client
+	stop   chan struct{}
+}
+
+// NewController creates a Controller.
+func NewController(loadBalancer LoadBalancer, kubernetesClient k8s.Client) Controller {
+	return &controller{
+		lb:     loadBalancer,
+		client: kubernetesClient,
+		stop:   make(chan struct{}),
+	}
+}
+
+func (c *controller) Run() {
+	log.Infof("Starting controller for %v and %v", c.lb, c.client)
+
+	watcher := k8s.NewWatcher()
+	defer close(watcher.Done())
+
+	err := c.client.WatchIngresses(watcher)
+	if err != nil {
+		log.Fatalf("Unable to watch ingresses: %v", err)
+		c.Stop()
+	} else {
+		go c.watchForUpdates(watcher)
+	}
+
+	// initialize with current state of ingress
+	c.updateLoadBalancer()
+
+	<-c.stop
+	log.Infof("Controller has stopped")
+}
+
+func (c *controller) watchForUpdates(watcher k8s.Watcher) {
+	for {
+		select {
+		case <-c.stop:
+			return
+		case update := <-watcher.Updates():
+			if val, ok := update.(k8s.Ingress); ok {
```

Comment:
Can we just type the channel rather than have a generic watcher? Having maintained unyped actor systems this scares me :)